### ISSUE TITLE
TypeScript 5.0 Iteration Plan

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#51362]
This document outlines our focused tasks for TypeScript 5.0. It minimally indicates intent to investigate tasks or contribute to an implementation. Nothing is set in stone, but we will strive to complete these tasks in a reasonable timeframe.

Date           | Event
---------------|-------------------------
November 15th  | TypeScript 4.9 Release
January 20th   | Create 5.0 Beta (5.0.0) Build for Testing
January 24th   | **TypeScript 5.0 Beta Release**
February 24th  | Create 5.0 RC (5.0.1) Build for Testing
February 28th  | **TypeScript 5.0 RC Release**
March 10th     | Create 5.0 Final (5.0.2) Build for Testing
March 16th     | **TypeScript 5.0 Final Release** 🚀

```mermaid
gantt
    dateFormat  YYYY-MM-DD
    TypeScript 4.9 Stabilization Period : 2022-10-28, 2022-11-11
    TypeScript 5.0 Beta Development : 2022-10-28, 2023-01-19
    TypeScript 5.0 RC Development : 2023-01-19, 2023-02-24
    TypeScript 5.0 Stabilization Period : 2023-02-24, 2023-03-16
todayMarker stroke-width:5px,stroke:#0f0,opacity:0.5
```

## Language and Compiler Features

* [ECMAScript Decorators](https://github.com/microsoft/TypeScript/issues/48885)
* [Unify All `enum`s as Unions](https://github.com/microsoft/TypeScript/pull/50528)
* [Legacy Flag Deprecation](https://github.com/microsoft/TypeScript/issues/51000)
* [Investigate Bundler-Oriented Module Resolution Options](https://github.com/microsoft/TypeScript/issues/50152)
* [Support `.ts` as a Module Specifier for Bundler/Loader Scenarios](https://github.com/microsoft/TypeScript/issues/37582)
* [Declaration Files for Non-JavaScript Files](https://github.com/microsoft/TypeScript/issues/50133)
* [Allow Arrays in `extends` for `tsconfig.json` Files](https://github.com/microsoft/TypeScript/pull/50403)
* [Investigate Per-File Settings](https://github.com/microsoft/TypeScript/pull/49886)
* [`lib.d.ts` Updates](https://github.com/microsoft/TypeScript/issues/51361)
* [Make `@types/web` Versionable](https://github.com/microsoft/TypeScript-DOM-lib-generator/issues/1383)

## Editor Productivity

* [`switch`/`case` Snippet Completions](https://github.com/microsoft/TypeScript/pull/50996)
* [Investigate `SharedArrayBuffer`-Backed Host for Web Editing Contexts](https://github.com/sandersn/vscode-wasm-typescript/)
* [Investigate Limited ATA-Like Scenarios in Web Editing Contexts](https://github.com/microsoft/vscode/issues/172887)
* [Implement Cancellation in Web Editing Contexts](https://github.com/microsoft/vscode/issues/148967)
* [Easy Run/Debug for Loose TypeScript Files](https://github.com/microsoft/vscode/issues/124280)

## Performance

* [Investigate Module Resolution Caching in `.tsbuildinfo` Files](https://github.com/microsoft/TypeScript/pull/41004)
* [Optimizing `.tsbuildinfo`](https://github.com/microsoft/TypeScript/pull/51246)
* [Investigate Tooling for JIT Deoptimizations](https://github.com/microsoft/TypeScript/issues/44211)

## Infrastructure

* [Migrate TypeScript to Modules](https://github.com/microsoft/TypeScript/issues/35210)
* [Enable `strictFunctionTypes` on TypeScript Codebase](https://github.com/microsoft/TypeScript/pull/49929)

## Website

* [Make Onboarding and Cross-Platform Installation Smoother](https://github.com/microsoft/TypeScript-Website/issues/2500)

**Comment by DanielRosenwasser** [comment]
The 4.9 iteration plan is over at https://github.com/microsoft/TypeScript/issues/50457.

**Comment by akononenko-sumo** [comment]
It's so sad that adding proper type support for `throw` is not on the roadmap for v5...

**Comment by MLoughry** [comment]
One idea I had previously floated to @amcasey was to add compiler/server APIs with asynchronous signatures in 5.0 (even if they're still synchronous under the hood), and deprecate the synchronous APIs. This would allow Typescript to later (6.0?) remove the synchronous APIs and unblock the ability to perform asynchronous I/O (or even worker-based parallelization?) in future versions.

**Comment by lukeapage** [comment]
I’d love to see the core team engage with issues that are for example 400+ thumbs up and marked “awaiting feedback” rather than issues that seem less important.
Even if the answer is that the team consider it to not be typescript’s responsibility and close the issue - i would rather that than limbo and newer but less important features being tackled.

https://github.com/microsoft/TypeScript/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc+label%3A%22Awaiting+More+Feedback%22

**Comment by patroza** [comment]
@lukeapage I smell another 400 thumbs up incoming..

**Comment by rudy-xhd** [comment]
https://github.com/microsoft/TypeScript/issues/10571 (Allow skipping generics) would be highly appreciated

**Comment by mindlid** [comment]
Any plan to add experimental support [Operator Overloading](https://github.com/tc39/proposal-operator-overloading), [Pipeline Operator](https://github.com/tc39/proposal-pipeline-operator) or extension methods?

**Comment by ljharb** [comment]
@mindlid It would be highly inappropriate to add support for any proposal prior to stage 3.

**Comment by mindlid** [comment]
@ljharb any feature can be added despite being a proposal or not, that's the point of being a superset of javascript

**Comment by ljharb** [comment]
@mindlid *can* be, sure. but a) it's not a superset in any way, and b) proposals aren't in JavaScript until they're stage 4, so your argument doesn't hold up anyways.

**Comment by mindlid** [comment]
> a) it's not a superset in any way
hmm... aren't features like string literate templates or enums not superset features???? Any feature in TS which is not in JS falls into the SuperSet, including the type system, it's basic set logic.

> b) proposals aren't in JavaScript until they're stage 4
okay, don't be confused, who is talking about JS?, Typescript is not limited to the features of Javascript, JS it's only a subset of the TS superset.
According to your logic, you may want to remove the type system from Typescript until [Proposal Type Annotations](https://github.com/tc39/proposal-type-annotations) gets into Javascript

> It would be highly inappropriate to add support for any proposal prior to stage 3.
Features like [Decorators](https://www.typescriptlang.org/docs/handbook/decorators.html) have been in experimental support before Stage 3, that why you run them with `--experimental<featureName>`, and that's why the `--experimental<featureName>` parameters exist

TS is not just JS with types, is it's own set of unique features plus the ones of JS

@ljharb

**Comment by mindlid** [comment]
> a) it's not a superset in any way

hmm... aren't features like string literate templates or enums not superset features???? Any feature in TS which is not in JS falls into the SuperSet, including the type system, it's basic set logic.

> b) proposals aren't in JavaScript until they're stage 4

okay, don't be confused, who is talking about JS?, Typescript is not limited to the features of Javascript, JS it's only a subset of the TS superset.
According to your logic, you may want to remove the type system from Typescript until [Proposal Type Annotations](https://github.com/tc39/proposal-type-annotations) gets into Javascript

> It would be highly inappropriate to add support for any proposal prior to stage 3.

Features like [Decorators](https://www.typescriptlang.org/docs/handbook/decorators.html) have been in experimental support before Stage 3, that why you run them with `--experimental<featureName>`, and that's why the `--experimental<featureName>` parameters exist

TS is not just JS with types, is it's own set of unique features plus the ones of JS

@ljharb

**Comment by ljharb** [comment]
@mindlid yes, and TS's inclusion-by-default of a five-times-obsolete form of decorators is one of the larger mistakes it's made, as I believe they've acknowledged. I'm well aware TS is not just JS with types - that's why it's not a superset.

The issue is that pre-stage-3 features are subject to change a lot, especially the ones you mentioned - the possible churn isn't worth it.

**Comment by mindlid** [comment]
you cannot call a mistake to change a experimental feature, that's the point of being marked as experimental, users can test, give feedback, improve the design, typescript is used as guinea pig for some future ECMAScript features, so thanks Typescript for your  "failures", so the final ECMAScript feature will learn from your errors.

On the other hand if you use an experimental feature treating it as stable, and complaining about breaking changes, that's not a Typescript mistake, but a clear misunderstanding of the purpose.

**Comment by mindlid** [comment]
you cannot call a mistake to change a experimental feature, that's the point of being marked as experimental, users can test, give feedback, improve the design, typescript is used as guinea pig for some future ECMAScript features, so thanks Typescript for your "failures", so the final ECMAScript feature will learn from your errors.

On the other hand if you use an experimental feature treating it as stable, and complaining about breaking changes, that's not a Typescript mistake, but a clear misunderstanding of the purpose.

@ljharb

**Comment by ljharb** [comment]
@mindlid I'm going to stop arguing about this here since the TS team has stated their policy on this in the past, and it's not worth discussing unless they change it :-) TC39 has lots of ways to get feedback on syntax features prior to stage 3, and "ship them in TS" needn't be one of them.

TS team, I urge you to hide these comments since they will likely be distracting to those attempting to read the issue.

**Comment by dilame** [comment]
Dear TS team, i really appreciate and respect your excellent work, but so much people are waiting for the `throws` statement, why isn't it in the 5.0 roadmap? FMPOV it's the this functionality is directly related to the purpose of the existence of the TypeScript – typed JS. This is the feature i really miss more then any other feature – it lacks type-safety a lot. 
Would be very grateful if some of team members could give a reply about it

**Comment by RyanCavanaugh** [comment]
I think we'll probably take a post-5.0 activity of "Saying 'no' to a bunch of things that aren't going to happen in the next five years" (I can't wait for the flood of 👎s 😀 that's going to get me).

I would request that people not use the 5.0 iteration plan any further for pointing out that our suggestion backlog exists.

**Comment by Erik1001** [comment]
Allowing .ts extension would enable code sharing with a Deno backend.
+1 for "Support .ts as a Module Specifier for Bundler/Loader Scenarios".

**Comment by RobertAKARobin** [comment]
+1 for the `.ts` extension. This would be huge.

**Comment by polkovnikov-ph** [comment]
> "Saying 'no' to a bunch of things that aren't going to happen in the next five years"

👍 for audacity to think TypeScript will survive 5 years with such neglect to community opinion and codebase quality. That's the spirit.

![image](https://user-images.githubusercontent.com/2983969/208020649-0a284f56-5942-4148-8da3-aae3ab415b17.png)

**Comment by RyanCavanaugh** [comment]
What's that even a graph of?

**Comment by RyanCavanaugh** [comment]
Ah so it's a graph of commit count. Two people who would commit on basically every keystroke (they were good engineers, it was just their style) were active over 2015-2016 and 2017-2019, and we switched to squashing commits on merge around 2020 IIRC. Commit count is usually not the best measure of a project's, well, anything, since it entirely depends on arbitrary things like squashing and how often the people working on the project like to do intermediary commits. Of all the things that worry me about TS's future, not having enough git objects is the least of my concerns 😅

**Comment by dilame** [comment]
> Of all the things that worry me about TS's future

Could you please share your thoughts about these things? Would be very interesting to read!

**Comment by SmashingQuasar** [comment]
Great to see TypeScript evolve even further, thanks for the time invested in this project!

With that being said, I am a little bit surprised as this hardly seems like a major release. Most of these changes are relatively niche and hardly breaking changes.

I understand this is not an issue to post suggestions so I won't. I just want to express that this release seems disconnected from the current TypeScript "issues". For example, having an option is `tsconfig.json` to force all thrown to be instances of `Error` and thus make `error` not be `unknown` but `Error` in `try..catch` statements seems like a much more important change than unifying Enums or catering to bundlers. Actually I am surprised than catering to bundlers is within the scope of TypeScript. It seems like something that should be it's own separate thing.

Anyway, I hope we get to see more impactful changes in 5.x versions!

**Comment by Jakob5358** [comment]
> With that being said, I am a little bit surprised as this hardly seems like a major release. Most of these changes are relatively niche and hardly breaking changes.

I think typescript doesn't use semver because marketing or smth, meaning that 5.0 and 4.4 aren't that different, the versioning is just a way to know if you're on a newer or older version.

**Comment by SmashingQuasar** [comment]
> I think typescript doesn't use semver because marketing or smth, meaning that for example 5.0 and 4.9 aren't that different, the versioning is just a way to know if you're on a newer or older version.

Can anyone confirm or infirm this please?
I'd be interested to know if TypeScript is following a specific logic regarding releases if not using SemVer.
Versioning cannot be marketing, especially for a technology aimed exclusively at IT professionals. We need version numbers to be reliable and have a clear contract on how they behave.

**Comment by SmashingQuasar** [comment]
@rubiesonthesky thanks for the link!
This is going off-topic so I won't keep on going with this discussion but I appreciate the effort you put into your researches! 👍 (I also want to point out that not using SemVer and relying on x.0 => x.9 then x+1.0 => x+1.9 is absolutely terrible in my opinion)

**Comment by ssalbdivad** [comment]
Every new TypeScript version has breaking changes are listed at the end of its release notes. They also tend to affect a very small number of users and/or be correcting behavior that is obviously wrong. When they do require updates, they tend to limited in scope.

For a project like this, SemVer doesn't really add much value.

**Comment by Zain-ul-din** [comment]
@RyanCavanaugh Hey, Do you have any plan to improve `OOP` in `typescript 5.0`

**Comment by danielrentz** [comment]
The [roadmap page](https://github.com/microsoft/TypeScript/wiki/Roadmap) is a bit outdated
- link list to roadmap issues ends at Jun 2021
- 4.9 features are not completed
- planned 5.0 features are missing

**Comment by abdatta** [comment]
I see some negativity in the comments, so I just want to add I'm extremely grateful to the TS team for Typescript, completely revolutionizing the experience of working with JS on large projects. Like all big projects, this too has some pitfalls and some highly requested features that are yet to make it in. I hope we reach there some day as well, but for now the 5.0 features are still very exciting! Looking forward to it ❤️ (Particularly excited about the per file settings, as it will help in enabling strict setting gradually in large repos).

**Comment by DanielRosenwasser** [comment]
@danielrentz thanks for reminding me. The feature roadmap should be up to date, link back to iteration plans across 4.8 to 5.0, and will more accurately reflect our work for 5.0 at this point than the above iteration plan.

**Comment by daitay4** [comment]
I just love how every second (average at best) beginner who never even took part in any large scale project using vanilla JS just spew negative bu****it as if they know better or even have a concrete suggestion as for how to improve on the current state.

Anyways, for me one of the things I'd really love to see is performance improvements, though I understand that's probably not gonna happen anytime soon.

Keep up the good work and ignore the haters, this project single-handedly improved web design soooo much

**Comment by polkovnikov-ph** [comment]
> know better or even have a concrete suggestion

As a person who has been using TypeScript since its public beta, I have quite a few things to say about the progress of the project. Mainly, I'm used to stress that ad hoc development of type systems with no regard to the last 100 years of research in type theory unsurprisingly leads to major bugs. Let's take a glimpse on what it means in this _completely correct_ TS program:

```typescript
// oops! type of JSON.stringify is broken
{
    const a = JSON.stringify(undefined);
    a.toString(); // Cannot read properties of undefined
}

// oops! type inference for overloads is broken
{
    const d = document.createElement('div');
    d.className = 'test';
    document.body.appendChild(d);
    const r: null | HTMLScriptElement = document.querySelector('.test');
    r?.src.toString(); // Cannot read properties of undefined
}

// oops! type system doesn't handle non-enumerable properties
{
    const rect1 = document.body.getBoundingClientRect();
    const rect2 = {...rect1};
    console.log(rect2.x.toString()); // Cannot read properties of undefined
}

// oops! not even properties from prototypes
{
    const a = new Map();
    const b = {...a};
    b.get(1); // b.get is not a function
}

// oops! narrowing is broken
{
    const m: { [key: string]: string } = {};
    const i = '123';
    const j: string | null = m[i];
    j.toString(); // Cannot read properties of undefined
}

// oops! narrowing is broken again
{
    type A =
        | {a: 0, b: undefined}
        | {a: 1, b: string}
    ((x: A) => {
        x.a = 1;
        if (x.a) {
            x.b.toString(); // Cannot read properties of undefined
        }
    })({a: 0, b: undefined});
}

// oops! and again narrowing is broken
{
    ((x?: string) => {
        if (typeof x === 'string') {
            const fn = () => x = undefined;
            fn();
            console.log(x.toString()); // Cannot read properties of undefined 
        }
    })('a');
}

// oops! what is this again? broken narrowing?
{
    let cnt = 0;
    const x = {
        get a(): number | null {
            if (cnt === 0) {
                cnt = 1;
                return 1;
            } else {
                return null;
            }
        }
    };
    if (x.a) {
        x.a.toString();
    }
}

// oops! enums are broken
{
    enum A { B }
    ((x: A) => [0][x].toString())(10); // Cannot read properties of undefined
}

// oops! method calls are broken
{
    class A {
        private a = 'a';
        b() { return this.a.toString(); }
    }
    const {b} = new A();
    b(); // Cannot read properties of undefined
}

// oops! subtyping is broken
{
    type A = { a(): string };
    type B = { b(): string };
    const f = (x: A | B) => "a" in x ? x.a() : x.b();
    const x = { a: 10, b() { return "hello"; } };
    const y: B = x;
    f(y); // x.a is not a function
}

// oops! subtyping is broken again
{
    const a = (b: { c(d: {}): void }) => b.c({e: 1});
    a({ c: (d: { f: 1 }) => d.f.toString() }); // Cannot read properties of undefined
}

// oops! type system doesn't even handle TDZ
{
    const f = () => x;
    f(); // Cannot access 'x' before initialization 
    const x = 0;
}
```

The top-level suggestion is, of course, to start fixing major bugs in type system, like
- do binding group analysis properly;
- make a distinction between open and closed object types (and type the `in` operator in a way that narrows on either of its arguments);
- handle union and intersection types symmetrically (for example, fix the intentional bug that handles the last overload);
- at least try to create some introduction and elimination syntax forms for the types that don't have any (such as mapped object types);
- decide on whether recursive types use specialization, or if they are proper mu-types, and handle them consistently.

A good starting point would be to
- open EcmaScript standard, find mentions of *Error, and write tests that fail in that way
- write a fuzzer that finds cases when code doesn't use any unsafe features (`as`, `any`, type guards, overloads etc.) and TS doesn't find any bugs, but it fails in runtime, and add those to a set of tests too
- use number of failing tests as KPI of TS team

**Comment by RyanCavanaugh** [comment]
> make developers fix bugs instead of adding new features

Things you don't like about software aren't bugs. 100% soundness has been a documented non-goal of the project for 7+ years. Pretty much everything you listed there is either entirely by-design (if these are your goals for a language, literally please do use PureScript or Hegel instead), is already implemented (you forgot to turn on `noUncheckedIndexedAccess`!), or is a tracked *feature request* to seal up a known unsoundness.

**Comment by polkovnikov-ph** [comment]
Sorry, but this is some kind of misdirection. "Soundness" is merely a polite way to say "lack of bugs" in academia. If the nongoal is expressed in plain language (such as "we don't intend to write a compiler that works"), it makes much less sense. Even if there is trade-off between handling all the bugs and type system complexity, and it's explicitly specified, the language is still said to be sound (for example, Haskell's type system is sound if you don't do things such as `unsafePerformIO`).

Unfortunately, TS doesn't have any specification, not even informal, yet alone specification of its trade-offs, and is genuinely unsound, i.e. has bugs in basic functionality such as array methods (not even type of `reduce` is correct!), JSON methods, objects and function calls. The specified examples are just a minified form of the problems people encountered in production. If that's "by design", please, reconsider what the design is.

**Comment by evelant** [comment]
It strikes me that since ts has become so ubiquitous and gained so many advanced features there are seemingly two distinct categories of users. 

On one hand there are people who want typescript per its original design goals, primarily js with erasable types for incremental adoption.

On the other hand there are people for whom plain js is more or less a compile target, they never write it directly. I think it's this class of users who sometimes are critical of TS because it's close to their use case of an advanced strictly typed language that compiles to optimized js but they forget that isn't the design goal of TS.

I think there's probably room for another language in the ecosystem. A sibling of TS with type directed output, nominal types, soundness, and other advanced features outside of the original ts design goals. That is IMO what a lot of TS detractors want TS to be, hence the frustration. IDK if such a thing could live under the TS umbrella or if it need to come to life as its own language.

Perhaps some extra documentation or prominent featuring of the TS design goals could help people realize that TS isn't intended to be exactly what they're looking for. Might reduce the amount of frustrated comments like the above.

**Comment by RyanCavanaugh** [comment]
People wishing to rehash old discussions can pick up at https://github.com/microsoft/TypeScript/issues/9825#issuecomment-351107584 to see where this is going to go; let's keep further discussion here on-topic.

**Comment by daitay4** [comment]
@polkovnikov-ph  So many of given broken examples would be avoided by simply working in a "more correct" way, i.e. avoid mutations when possible.

In my experience simply applying some stricter linting avoids many of said cavities. Now whether or not this should fall under Typescript's umbrella is a different discussion, as already mentioned in this thread. The way I see it - the answer is no, but I guess that's debatable.

All of your given examples are not real-world case, and if anyone submits code for me to review which looks anything like any one of your examples I'd send them back to rewrite in a heartbeat.

**Comment by Thundercraft5** [comment]
Could **[this issue](#52035)** be investigated? It's a small but important and far-reaching bug for HKT and FP users, if fixed will allow some basic HKT in TS for the **first time**.

**Comment by DanielRosenwasser** [comment]
We're going to try to get the beta out tomorrow. We've had to do some manual testing, and documenting's taken a bit longer.

**Comment by Jack-Works** [comment]
@storybook+react-docgen-typescript-plugin@1.0.2-canary.6.9d540b91e815f8fc2f8829189deb00553559ff63.0 breaks with TS 5.0-beta.

> TypeError: typescript_1.default.createIdentifier is not a function

I know `ts.createIdentifier` is deprecated (`ts.factory.create*` instead). Are those deprecated methods totally removed in TS 5.0?

**Comment by jakebailey** [comment]
Yes, see #52314. They gave been deprecated for the past 2+ years (since 4.0).

**Comment by Jack-Works** [comment]
Thanks, I'll open an issue in that library. But this breaking change is not documented in https://github.com/microsoft/TypeScript/wiki/API-Breaking-Changes

**Comment by jakebailey** [comment]
> Thanks, I'll open an issue in that library. But this breaking change is not documented in [microsoft/TypeScript/wiki/API-Breaking-Changes](https://github.com/microsoft/TypeScript/wiki/API-Breaking-Changes)

I'll document it there, though I'm pretty sure that use of these functions should have been logging deprecation warnings since TS 4.1.

**Comment by AliMD** [comment]
> It's so sad that adding proper type support for `throw` is not on the roadmap for v5...

This is not version `5`! its just version `4.10` or better say version `3.20`! Microsoft does not understand semantic versioning and always make jump major-version after `x.9`.
Because Microsoft developers do not have a proper understanding of `x.10 > x.9` and this is confusing for them.
If TypeScript wasn't open source, it would be version 2023 now 😆

**Comment by Jack-Works** [comment]
> If TypeScript wasn't open source, it would be version 2023 now!

Some OOT just want to notify you that some other open source languages are also versioned by year like ECMAScript 2022 or rust edition 2022 (well that's not a "version" strictly)

**Comment by gund** [comment]
> This is not version `5`! its just version `4.10` or better say version `3.20`!

There have been many breaking changes between v3 and v4 even between minors, so if anything it should've been version 6+ (probably 10ish) 😁

Also it's important what TS considers it's "public API" - if it's only usage of the compiler then it's one thing but if it also includes API to work with AST then there were even more breaking changes and the version should've been even higher by now (20+?).

Either way in the current state TS versioning is hardly following a SemVer and you should just treat it as "edition" which does not tell you anything about it's BC guarantees and your best friend is their release notes.

**Comment by RyanCavanaugh** [comment]
> This is not version 5! its just version 4.10 or better say version 3.20

There is [literally a long list of breaking changes which we specifically put in this release](https://github.com/microsoft/TypeScript/issues/51909).

People can have a lot of semver complaints but "5.0 was incorrectly a major bump" is not one of them.

**Comment by nstepien** [comment]
Would it be too late to add "change array by copy" [0] methods types in TS 5.0?
They shipped in Chrome 110 stable, and have reached stage 4 [1].
[0] https://github.com/tc39/proposal-change-array-by-copy
[1] https://github.com/tc39/proposals/commit/ad4df8435f27f39eda26db3b940ae151980c8015

**Comment by SmashingQuasar** [comment]
> > This is not version 5! its just version 4.10 or better say version 3.20
> 
> There is [literally a long list of breaking changes which we specifically put in this release](https://github.com/microsoft/TypeScript/issues/51909).
> 
> People can have a lot of semver complaints but "5.0 was incorrectly a major bump" is not one of them.

I don't want to start a debate about it, just want to point out that deprecations do not fall under the breaking change status because they don't break anything yet. Removal of syntax or features is what falls under the breaking changes nomenclature.

Overall, I think the question is not really "Have you introduced breaking changes in 5.0?" but more of "Shouldn't you use a real Semantic Versioning instead of doing x.1 => x.9 on every pseudo-major?".

**Comment by DanielRosenwasser** [comment]
The TypeScript 5.1 Iteration Plan is available [here](https://github.com/microsoft/TypeScript/issues/53031).

**Comment by DanielRosenwasser** [comment]
We will likely publish the RC tomorrow.

**Comment by RyanCavanaugh** [comment]
> I don't want to start a debate about it, just want to point out that deprecations do not fall under the breaking change status because they don't break anything yet. Removal of syntax or features is what falls under the breaking changes nomenclature.

They do break things, because your build is going to start failing until you opt into the "give me back the old behavior" flag.

**Comment by SmashingQuasar** [comment]
> > I don't want to start a debate about it, just want to point out that deprecations do not fall under the breaking change status because they don't break anything yet. Removal of syntax or features is what falls under the breaking changes nomenclature.
> 
> They do break things, because your build is going to start failing until you opt into the "give me back the old behavior" flag.

That's not a deprecation. A deprecation triggers a notice or a warning without behavior change. You are describing a retro-compatiblity option which is not the same at all. To me knowledge, namespaces are widely deprecated and are still working just fine without any optional flag. Optional flags are also used for opt-in to some more strict features. It's really not a good example of deprecation in my opinion. Again, let's not start a debate about that.

**Comment by RyanCavanaugh** [comment]
Namespaces aren't deprecated.

**Comment by SmashingQuasar** [comment]
My bad, I was actually think about this syntax:

```ts
module Foo { }
```
which still works without any flag but is deprecated in favor of namespaces. Thanks for pointing it out for others who may not know about it. This is a good example of deprecation done right by TypeScript. (it would benefit from triggering a warning which does not seem to be the case with TS 4.9.x)

**Comment by SmashingQuasar** [comment]
Yes, but just remember that this whole discussion was about TypeScript not following SemVer and instead using an arbitrary versioning that uses the SemVer format but not the logic. If you don't want to use SemVer, it would be preferable to simply use raw version numbers like some other software do such as TS 5, TS 6... TS 123.
Then, I pointed out that your link listed deprecations, which do not constitute a breaking change under the SemVer logic and should instead be listed within a minor version. You insisted that these were breaking changes and they are not. Overall, deprecations should not break existing code, otherwise they are not deprecations, they are simply feature suppression with possibly a compatibility bridge that needs manual activation.

**Comment by RyanCavanaugh** [comment]
It seems like maybe you're not fully aware of how these deprecations are being rolled out? If you currently are targeting ES3 and try to use TS5, you will get an error you didn't get before. That's exactly the situation (my build was passing and now it's not) that 95% of "This should have been a semver major bump" complaints are about. The ease of getting back to zero errors is, in all cases, wholly irrelevant.

**Comment by DanielRosenwasser** [comment]
@typescript-bot bump release-5.0

**Comment by SmashingQuasar** [comment]
> It seems like maybe you're not fully aware of how these deprecations are being rolled out? If you currently are targeting ES3 and try to use TS5, you will get an error you didn't get before. That's exactly the situation (my build was passing and now it's not) that 95% of "This should have been a semver major bump" complaints are about. The ease of getting back to zero errors is, in all cases, wholly irrelevant.

Again, deprecations should not break existing code. That is the whole concept of a deprecation. If they are breaking existing code, they are not deprecations, they are feature removal with a retro-compatibility bridge. If they are handled that way, then sure, those are breaking changes that constitute a major version. They should however be listed as such and not as deprecations.

**Comment by trim21** [comment]
> We're having an issue with a build pipeline, so we'll be targeting the stable release for Thursday, March 16th.

is this the same reason why 5.0.2 is not released?

**Comment by DanielRosenwasser** [comment]
Ugh, that is a typo in the release plans. 5.0.2 is the release I'm talking about.

**Comment by sacummings91** [comment]
It's crazy that people have the nerve to come in here and complain. TS is literally free for us to use and everyone on the team puts a massive amount of effort into it. You should just be grateful it exists.

**Comment by jez9999** [comment]
It's the 16th... any chance of TS5 today?

**Comment by RyanCavanaugh** [comment]
Yes! https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/

**Comment by mindlid** [comment]
@polkovnikov-ph https://civet.dev/ supports ES proposals and compiles to typescript

**Comment by DanielRosenwasser** [comment]
@typescript-bot bump release-5.0

**Comment by DanielRosenwasser** [comment]
@typescript-bot bump release-5.0

**Comment by dejayc** [comment]
> It's crazy that people have the nerve to come in here and complain. TS is literally free for us to use and everyone on the team puts a massive amount of effort into it. You should just be grateful it exists.

I agree.  On average, it only takes 3 to 5 years for a moderately popular feature request to be considered for implementation, and look at all the ideological purity we get in return!
